### PR TITLE
ext_proc: Ensure that timer is always cancelled

### PR DESCRIPTION
**Pull request by gbrail**
Commit Message: Ensure that the timer is cancelled when an "ImmediateResponse"
message is received.
Additional Description:
Risk Level: Low
Testing: New unit tests
Docs Changes:
Release Notes:
Platform Specific Features:
Signed-off-by: Gregory Brail <gregbrail@google.com>

### CODE REVIEW

**Comment by gbrail** [comment]
Presubmit tests are failing because some other modules have fallen below the coverage threshold -- the ext_proc code has fine coverage.

File: test/extensions/filters/http/ext_proc/filter_test.cc [hunk 1]
```diff
@@ -72,6 +86,15 @@ class HttpFilterTest : public testing::Test {
     request_headers_.setMethod("POST");
   }
 
+  void TearDown() override {
+    for (auto* t : timers_) {
+      // This will fail if any timer is un-disabled at the end of the test run.
```
Comment (snowp) [inline test/extensions/filters/http/ext_proc/filter_test.cc:91]: `is still enabled`? avoids double negative

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 2]
```diff
@@ -465,6 +467,9 @@ void Filter::onGrpcError(Grpc::Status::GrpcStatus status) {
 
 void Filter::onGrpcClose() {
   ENVOY_LOG(debug, "Received gRPC stream close");
+  if (processing_complete_) {
+    return;
+  }
```
Comment (snowp) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:472]: How is this related to disabling timers?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, a couple comments but I think this makes sense

**Review by cboitel — COMMENTED** [review]
In other onXXX, test is performed after stats are updated not before. In most cases, metrics won't be updated making the closed stream useless.

File: test/extensions/filters/http/ext_proc/filter_test.cc (hunk 1 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/filter_test.cc:91]: Yep, simplifying this one.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 2 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:472]: This was a belt-and-suspenders check that it turns out we don't actually need, so I'm going to take it out.

**Review by snowp — APPROVED** [review]
Thanks!

/retest

**Comment by snowp** [comment]
Looks like a timeout in one of the ext_proc tests, can you take a look? Not sure if its this PR or if the test is just flaky

**Comment by gbrail** [comment]
I was trying to be efficient and changed the size of that test to "small" but since I see it timing out in 60 seconds it's clearly "medium" at least in the tsan case.

**Comment by cboitel** [comment]
Is there any chance to have this change in one of the v1.19 maintenance release ?
